File: packages/mlforge_sdk/mlforge_sdk/benchmark.py
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, ConfigDict, Field

from .http import HttpClient

class BenchmarkResult(BaseModel):
    model_config = ConfigDict(extra="ignore")
    id: str
    job_id: str
    metrics: Dict[str, Any]
    telemetry_summary: Dict[str, Any]
    created_at: Optional[str] = None
    model_id: Optional[str] = None
    dataset_id: Optional[str] = None
    task: Optional[str] = None
    framework: Optional[str] = None
    hardware: Optional[str] = None
    precision: Optional[str] = None


class BenchmarkJob(BaseModel):
    model_config = ConfigDict(extra="ignore")
    id: str
    model_id: str
    dataset_id: str
    task: str
    framework: str
    hardware: str
    precision: str
    batch_size: int
    status: str
    progress: float = 0.0
    logs: List[str] = Field(default_factory=list)


class BenchmarkRunResponse(BaseModel):
    model_config = ConfigDict(extra="ignore")
    job_id: str
    status: str
    message: str

class BenchmarkClient:
    def __init__(self, http: HttpClient):
        self._http = http

    def list_results(self) -> List[BenchmarkResult]:
        data = self._http.get("/benchmark/results/all", params={"limit": 100})
        return [BenchmarkResult(**r) for r in data]

    def list_jobs(
        self,
        *,
        status: Optional[str] = None,
        model_id: Optional[str] = None,
        limit: int = 100,
    ) -> List[BenchmarkJob]:
        params: Dict[str, Any] = {"status": status, "model_id": model_id, "limit": limit}
        data = self._http.get("/benchmark/jobs", params=params)
        return [BenchmarkJob(**j) for j in data]

    def run(self, ctx: Dict[str, Any]) -> BenchmarkRunResponse:
        data = self._http.post("/benchmark/run", json_body=ctx)
        return BenchmarkRunResponse(**data)

    def validate(self, ctx: Dict[str, Any]) -> Dict[str, Any]:
        return self._http.post("/benchmark/validate", json_body=ctx)

    def get_job(self, job_id: str) -> BenchmarkJob:
        data = self._http.get(f"/benchmark/{job_id}")
        return BenchmarkJob(**data)

    def get_result(self, job_id: str) -> BenchmarkResult:
        data = self._http.get(f"/benchmark/{job_id}/result")
        return BenchmarkResult(**data)

File: packages/mlforge_js_sdk/src/benchmark.ts
import type { HttpClient } from "./http.js";

export interface BenchmarkResult {
  id: string;
  job_id: string;
  metrics: Record<string, unknown>;
  telemetry_summary: Record<string, unknown>;
  created_at?: string;
  model_id?: string;
  dataset_id?: string;
  task?: string;
  framework?: string;
  hardware?: string;
  precision?: string;
}

export interface BenchmarkJob {
  id: string;
  model_id: string;
  dataset_id: string;
  task: string;
  framework: string;
  hardware: string;
  precision: string;
  batch_size: number;
  status: string;
  progress?: number;
  logs?: string[];
}

export interface BenchmarkRunOptions {
  project_id: string;
  model_id: string;
  dataset_id?: string;
  task?: string;
  framework?: string;
  hardware?: string;
  precision?: string;
  batch_size?: number;
  [key: string]: unknown;
}

export class BenchmarkClient {
  constructor(private http: HttpClient) {}

  async listResults(opts?: { limit?: number }): Promise<BenchmarkResult[]> {
    return this.http.get<BenchmarkResult[]>("/benchmark/results/all", {
      limit: opts?.limit ?? 100,
    });
  }

  async listJobs(opts?: { status?: string; model_id?: string; limit?: number }): Promise<BenchmarkJob[]> {
    return this.http.get<BenchmarkJob[]>("/benchmark/jobs", {
      status: opts?.status,
      model_id: opts?.model_id,
      limit: opts?.limit ?? 100,
    } as Record<string, string | number | boolean>);
  }

  async run(opts: BenchmarkRunOptions): Promise<{ job_id: string; status: string }> {
    return this.http.post<{ job_id: string; status: string }>("/benchmark/run", opts);
  }

  async validate(opts: Record<string, unknown>): Promise<Record<string, unknown>> {
    return this.http.post<Record<string, unknown>>("/benchmark/validate", opts);
  }

  async getJob(jobId: string): Promise<BenchmarkJob> {
    return this.http.get<BenchmarkJob>(`/benchmark/${jobId}`);
  }

  async getResult(jobId: string): Promise<BenchmarkResult> {
    return this.http.get<BenchmarkResult>(`/benchmark/${jobId}/result`);
  }
}
